### PR TITLE
Feature/ml 523 moving tensor constructions to happen on graph compile

### PR DESCRIPTION
**Pull request by baykaner**
- rework compilation  so that tensors are initialised at compile time (not layer add time)
- removed graph auto-compilation in some cases
- graph shape auto-deduction now fully separate from tensor initialisation for layer

### CODE REVIEW

File: libs/ml/include/ml/ops/ops.hpp [hunk 1]
```diff
@@ -155,9 +155,14 @@ class Ops
 
   /// Should be called after shape linking in Graph to complete all initialisations, that depends
   /// on layer shapes (like trainable parameter tensors init. and so on)
-  virtual void CompleteConstruction()
+  virtual void CompleteShapeDeduction()
   {
-    // Empty deafult implementation for non-trainable Ops.
+    // Empty defualt implementation for Ops without shape
```
Comment (evsmithx) [inline libs/ml/include/ml/ops/ops.hpp]: spelling

File: libs/ml/src/layers/fully_connected.cpp [hunk 2]
```diff
@@ -132,21 +140,39 @@ void FullyConnected<TensorType>::CompleteConstruction()
...
+  //
+  //  // Construct weights and bias tensors
+  //  TensorType weights_data(weights_shape);
+  //  fetch::ml::ops::Weights<TensorType>::Initialise(weights_data, total_inputs_, total_outputs_,
+  //  init_mode_); TensorType bias_data = TensorType(this->batch_output_shape_);
+  //
+  //  this->SetInput(weights_name_, weights_data);
+  //  this->SetInput(bias_name_, bias_data);
+  //
+  //  this->Compile();
```
Comment (evsmithx) [inline libs/ml/src/layers/fully_connected.cpp]: Commented code

Comment (MissingNO57) [inline libs/ml/src/layers/fully_connected.cpp]: Commented out code

**Review by MissingNO57 — CHANGES_REQUESTED** [review]

File: libs/ml/src/layers/fully_connected.cpp [hunk 3]
```diff
@@ -87,28 +87,36 @@ FullyConnected<TensorType>::FullyConnected(SizeType in, SizeType out,
...
     this->ComputeBatchOutputShape(this->batch_input_shapes_);
-    CompleteConstruction();
+    CompleteShapeDeduction();
   }
 }
 
+/**
+ *
+ * @tparam TensorType
+ */
```
Comment (evsmithx) [inline libs/ml/src/layers/fully_connected.cpp:97]: Empty docstring

File: libs/vm-modules/tests/unit/ml/ml_model_tests.cpp [hunk 4]
```diff
@@ -412,26 +474,26 @@ TEST_F(VMModelTests, model_add_dense_relu)
...
-                       IGNORE_CHARGE_ESTIMATION);
+  TestValidFourDimLayerAdding(R"(model.add("conv2d", 10u64, 10u64, 10u64, 10u64);)",
+                              IGNORE_CHARGE_ESTIMATION);
 }
 
 TEST_F(VMModelTests, model_add_conv2d_relu)
 {
-  TestValidLayerAdding(R"(model.add("conv2d", 10u64, 10u64, 10u64, 10u64, "relu");)",
-                       IGNORE_CHARGE_ESTIMATION);
+  TestValidFourDimLayerAdding(R"(model.add("conv2d", 10u64, 10u64, 10u64, 10u64, "relu");)",
```
Comment (evsmithx) [inline libs/vm-modules/tests/unit/ml/ml_model_tests.cpp:495]: I'm not sure how these tests relate to the rest of the PR, but they're useful to have!

**Review by evsmithx — CHANGES_REQUESTED** [review]

File: libs/vm-modules/tests/unit/ml/ml_model_tests.cpp (hunk 4 above)
Comment (baykaner) [reply libs/vm-modules/tests/unit/ml/ml_model_tests.cpp:495]: previously it wasn't necessary to compile model to test adding the layers, as a result a little bit of test fixture editing was needed; other than that it's just a re-branding of an existing test

**Review by evsmithx — APPROVED** [review]

**Review by MissingNO57 — APPROVED** [review]
LGTM

**Pull request merged**